### ISSUE TITLE
In-world blocks do not implement interfaces from Block class

### ISSUE THREAD

**Issue by Dynious** [NOVA-Team/NOVA-Core#137]
```
<Dynious> My Block implements MyInterface, but this is false: if (world().getBlock(event.position).get() instanceof MyInterface)
<Dynious> This is either broken or bad design
<RX14-chibi> https://gist.github.com/RX14/dfd08d94bd328be6ff49
<Dynious> Yea, bad design
<RX14-chibi> look at vic's stuff, she must have done it Dynious 
<Dynious> I know, you´ll probably be able to do it using that, but this will confuse soooooo many people
<RX14-chibi> why?
<Dynious> Basically implementing an interface doesn´t actually implement the interface
<RX14-chibi> ehh
<Dynious> The block in world will not have implmented it
<Dynious> Whilest your class has
<RX14-chibi> uhh
<Dynious> I know why, but it´s stupid
<RX14-chibi> there must be a way
<Dynious> Yea, FMP does it
<RX14-chibi> then do it
<Dynious> Just gets all the interfaces and methods and copies them to the Block
```
Conversation from IRC. I know the reason behind this (we have BWBlocks), but this needs to be changed (or if there's a proper other way VERY well documented!). Modders that do not have a good understanding of the reasons behind this and are not experienced coders will be very confused and frustrated. When implementing something you expect your Block to actually implement it in-word. If this is a design decision (maybe you prefer only having implementations in components), World::getBlock should not return the same kind of class (Block) as the one the modder is working with, as they will expect it to have the same properties.

**Comment by anti344** [comment]
I vote for FMP-like stuff, but it'll be not easy to implement.

**Comment by Victorious3** [comment]
Is it needed? If so, same pass-through as for TileEntities.

**Comment by Dynious** [comment]
I don't mean on the Minecraft Block, but on the NOVA Block class. When you now implement anything on the Block class in NOVA it will not be on the MC version of the TE (and not on the BWBlock). This will be confusing if not properly documented. The whole 'component implements TE interface' is not straightforward and should be explained.

**Comment by Victorious3** [comment]
@Dynious Remember that we don't have an @Optional.Interface annotation. Putting the interfaces directly on the block would cause compatibility issues.

**Comment by Dynious** [comment]
I did not realize that, and this seems like a good argument to do things the way it's done now, but then getBlock should not return a Block instance, as you would think it would return an instance of the Block you placed there.

**Comment by anti344** [comment]
Maybe add some kind of annotation for it then? And in Java 8 there was something neat with annotation lists

**Comment by RX14** [comment]
@anti344 yeah you can have repeating annotations.

**Comment by Victorious3** [comment]
@Dynious Well, this is mainly for one-way compatibility. But you are right, there should be a method to get the specific implementation of the block. 
So to sum this up:
- Passthrough interface for blocks
- Better documentation on said interfaces
- A way to get the native block

Anything else?

**Comment by gjgfuj** [comment]
You should return the nova block class if the block is written in nova?
Yknow? It makes sense.

**Comment by Victorious3** [comment]
@gjgfuj Yeah, but this will always be a the NOVA block / the wrapped block. You wouldn't be able to do anything with the interfaces provided. Imagine an energy system for example where you have to check the neighbor blocks for an interface in order to receive input. (Bad example because bad design, but you get what I mean)

**Comment by Dynious** [comment]
With the points you give I'm not entirely sure if you get the issue I'm seeing. All features that are needed are already in NOVA, they are just made in a way that will be very confusing for new modders.

1. I make a Block class that implements MyInterface (or IDK a CoFH RF interface)
2. I place my Block class that implements MyInterface at a position
3. I check if the block at that position implements MyInterface.
4. This is always false, even though I do implement the interface on that block and I do get a Block instance.

What you need to do is add MyInterface to a component, then check for every component added to the block if it implements MyInterface. I would NOT find this out myself if I didn't check the NOVA/Wrapper source.

Suggested solution to make this more obvious: Do not return use Block.class to make a block and as a return for getBlock, as it's not the same thing. getBlock will return a block with the same components, but not the same class.

**Comment by RX14** [comment]
@Dynious people should be making their own components, not implementing interfaces.

**Comment by Dynious** [comment]
Exactly. This is different than MC, so it should actually be different and not some kind of weird merged mess of the two.

**Comment by Victorious3** [comment]
@Dynious You can't really force anybody to *not* use interfaces and do components for everything. Either they are consistent or they confuse themselves, I don't see what we could do about that.

**Comment by RX14** [comment]
Even though people shouldn't be using interfaces, it is confusing in the way @Dynious says.

**Comment by Dynious** [comment]
Yes, to explain why, here is a parody on the getBlock description. "Gets the Block on this position. But not actually the Block you put here, but some other block that acts like it's the block you put here. Will only have the same components, not the same interfaces or class. Basically, it's not at all what you put here, although we make it seem like it is."

**Comment by Victorious3** [comment]
@RX14 Oh they should. It's a part of the language we are using. Go for an interface if it makes sense, components don't magically remove all inheritance.

**Comment by calclavia** [comment]
As i mentioned , we should do dual passthroughs.

Components with passthroughs generates an implementation on the TE, that forwards to the block.
TE with interface generates a component (which implements said interface) that is added to BWBlock.

@Victorious3 Would you be able to do this?

**Comment by Victorious3** [comment]
@calclavia Maybe... But I still think it's wasted effort tbh. We already have an Object that implements said interface. Its also hardly cachable unless I transform TileEntity and add my own variable to it.

**Comment by calclavia** [comment]
@Victorious3 It's also about the design. Doing it your way we loose the entire well-designed-API feel. It's hacky to do it that way, and we don't want to expose hackiness into the Core itself.

**Comment by calclavia** [comment]
I agree with @Dynious's proposal: http://pastebin.com/DXCqtB8e

**Comment by Kubuxu** [comment]
My voice out of the grave:
```
In general we should discourage Interfaces on blocks.
It should be glue logic for existing APIs but we and modders should not use it to create new APIs.
```
PS: I will come back to NOVA. I took some detour via C99.

**Comment by calclavia** [comment]
@Kubuxu Agree
